README
nr_easy_workspace
=================

Provides a new workspace management frontend
that is easier to use than the original one
- it is based on pages, not single content elements.

It also features easy review email composition,
selectable reviewer and review message templates.



ChangeLog
2011-01-25 Alexander Opitz <alexander.opitz@netresearch.de>
* Set page into "review" state and not "publish". (12515)

2010-12-20 Alexander Opitz <alexander.opitz@netresearch.de>
* Version 0.5.2

2010-12-13 Alexander Opitz <alexander.opitz@netresearch.de>
* Fix preview link inside mails (partly respect language). (11022)
* Alternating rows in list view. (11023)

2010-10-25 Alexander Opitz <alexander.opitz@netresearch.de>
* Version 0.5.1
* Thanks for donations goes to
  - Delft University of Technology (Netherlands) http://tudelft.nl/ (9426, 9427,
    9428)
  - Universität Leipzig (Germany) http://uni-leipzig.de (ULS-65)

2010-10-21 Christian Weiske <christian.weiske@netresearch.de>
* Update documentation (NRFE-120)

2010-10-11 Alexander Opitz <alexander.opitz@netresearch.de>
* Don't show preview button on elements which are no page and no page_overlay,
  fix link for page_overlay (ULS-65)

2010-09-14 Alexander Opitz <alexander.opitz@netresearch.de>
* Rewrite preview to handle correct workspace. (9427)

2010-09-13 Alexander Opitz <alexander.opitz@netresearch.de>
* Possibility to have up to 99 Mail messages. (9426)
* Add /fileadmin/workspaceMessages/locallang.xml as a possible way to add self
  written messages (9426)
* Fix Domain to link of page preview in mail (9428)

2010-05-07 Alexander Opitz <alexander.opitz@netresearch.de>
* Version 0.5.0

2010-04-22 Alexander Opitz <alexander.opitz@netresearch.de>
* Resolve template loading problem, if typo3 directory is symlinked. (ULS-19)
* Release as 0.4.99. (ULS-19)

2010-04-22 Alexander Opitz <alexander.opitz@netresearch.de>
* Release as 0.4.98. (NRFE-7, NRFE-42)

2010-03-08 Alexander Opitz <alexander.opitz@netresearch.de>
* Add note about editors access to pages table. (NRFE-7, NRFE-42)

2010-03-05 Alexander Opitz <alexander.opitz@netresearch.de>
* Update documentation, add editor and reviewer setup informations (NRFE-7)

2010-01-05 Alexander Opitz <alexander.opitz@netresearch.de>
* Add support for "can live" and "rejected" to pagetree icon.
* Update documentation, add simple workspace setup

2009-09-28 Alexander Opitz <alexander.opitz@netresearch.de>
* change label_publish to label_action #5048

2009-09-07 Alexander Opitz <alexander.opitz@netresearch.de>
* Output uses typo3 templating
* Colorize state of review, also for "can live" and "rejected"
* fix pagetree icon
* multilanguage support for reviewer message through locallang.xml
* Hint if no reviewer is defined for workspace
* clean up dead and unused code
* Document functions and use PEAR coding standard

2009-08-24 Alexander Opitz <alexander.opitz@netresearch.de>
* Mark elements in Review, unset "in review" on main page if the page didn't changed.
* Update documentation
* Check workspace for pagetree icon
* Version 0.4.3

2009-08-19 Alexander Opitz <alexander.opitz@netresearch.de>
* Add administration entry about TSconfig var.

2009-08-19 Alexander Opitz <alexander.opitz@netresearch.de>
* Add Language id parameter to preview #3840
* Version 0.4.2

2009-08-17 Christian Weiske <christian.weiske@netresearch.de>
* Use proper API method to check if tree item is needed
* Version 0.4.1 (bug #4424)

2009-08-07 Christian Weiske <christian.weiske@netresearch.de>
* Add extra icon in tree view when page is under review
* Version 0.4.0 (bug #4424)

2009-08-04 Alexander Opitz <alexander.opitz@netresearch.de>
* New FAQ entry in documentation (Colors in tree view)

2009-08-04 Alexander Opitz <alexander.opitz@netresearch.de>
* class.tx_nreasyworkspace_tcemain.php nach UTF-8.
* Version 0.3.3

2009-07-28 Alexander Opitz <alexander.opitz@netresearch.de>
* Change eMail adress in ChangeLog
* Accept first Mountpoint as valid Mountpoint.
* Version 0.3.2

2009-07-06 Alexander Opitz <alexander.opitz@netresearch.de>
* Use $TYPO3_CONF_VARS['SYS']['sitename'] for messages.
* Version 0.3.1

2009-03-16 Dirk Förster <df@netresearch.de>
* correct documentation
* Version 0.3.0

2009-02-10 Alexander Opitz <alexander.opitz@netresearch.de>
* mark lines as "in review" if page is new and have changed sub elements
* set -f option for returnPath and use own mail function.
* fix usage of $this->BE_USER => $BE_USER

2009-01-26 Alexander Opitz <alexander.opitz@netresearch.de>
* translate description
* add documentation

2009-01-19 Alexander Opitz <alexander.opitz@netresearch.de>
* Version 0.2.0 with german messages.



ext_emconf.php
<?php

########################################################################
# Extension Manager/Repository config file for ext "nr_easy_workspace".
#
# Auto generated 20-12-2010 11:03
#
# Manual updates:
# Only the data in the array - everything else is removed by next
# writing. "version" and "dependencies" must not be touched!
########################################################################

$EM_CONF[$_EXTKEY] = array(
	'title' => 'Easy Workspace',
	'description' => 'An easy workspace administration tool for editors and reviewers.',
	'category' => 'be',
	'shy' => 0,
	'version' => '0.5.2',
	'dependencies' => '',
	'conflicts' => '',
	'priority' => '',
	'loadOrder' => '',
	'module' => '',
	'state' => 'beta',
	'uploadfolder' => 0,
	'createDirs' => '',
	'modify_tables' => '',
	'clearcacheonload' => 0,
	'lockType' => '',
	'author' => 'Netresearch',
	'author_email' => 'info@netresearch.de',
	'author_company' => '',
	'CGLcompliance' => '',
	'CGLcompliance_note' => '',
	'constraints' => array(
		'depends' => array(
			'typo3' => '4.2.1-4.4.99',
		),
		'conflicts' => array(
		),
		'suggests' => array(
		),
	),
	'_md5_values_when_last_written' => 'a:14:{s:9:"ChangeLog";s:4:"3fe0";s:6:"README";s:4:"ad0f";s:36:"class.tx_nreasyworkspace_tcemain.php";s:4:"bf4d";s:21:"ext_conf_template.txt";s:4:"f015";s:12:"ext_icon.gif";s:4:"a459";s:17:"ext_localconf.php";s:4:"f76b";s:14:"doc/manual.pdf";s:4:"5bbc";s:14:"doc/manual.sxw";s:4:"c3dc";s:22:"examples/locallang.xml";s:4:"20d1";s:38:"mod1/class.ux_SC_mod_user_ws_index.php";s:4:"aaaf";s:27:"mod1/class.ux_wslib_gui.php";s:4:"c2a1";s:30:"mod1/class.ux_wsol_preview.php";s:4:"0b61";s:18:"mod1/locallang.xml";s:4:"e130";s:17:"templates/ws.html";s:4:"2486";}',
	'suggests' => array(
	),
);

?>

ext_localconf.php
<?php
if (!defined ('TYPO3_MODE')) {
    die ('Access denied.');
}

$TYPO3_CONF_VARS[TYPO3_MODE]['XCLASS']['typo3/mod/user/ws/index.php']
    = t3lib_extMgm::extPath($_EXTKEY).'mod1/class.ux_SC_mod_user_ws_index.php';
$TYPO3_CONF_VARS[TYPO3_MODE]['XCLASS']['typo3/mod/user/ws/class.wslib_gui.php']
    = t3lib_extMgm::extPath($_EXTKEY).'mod1/class.ux_wslib_gui.php';
$TYPO3_CONF_VARS[TYPO3_MODE]['XCLASS']['typo3/mod/user/ws/wsol_preview.php']
    = t3lib_extMgm::extPath($_EXTKEY).'mod1/class.ux_wsol_preview.php';

$TYPO3_CONF_VARS['SC_OPTIONS']['t3lib/class.t3lib_tcemain.php']['processCmdmapClass']['nr_easy_workspace']
    = 'EXT:nr_easy_workspace/class.tx_nreasyworkspace_tcemain.php:tx_nreasyworkspace_tcemain';

$GLOBALS['TYPO3_CONF_VARS']['SC_OPTIONS']['GLOBAL']['recStatInfoHooks'][]
    = 'EXT:nr_easy_workspace/class.tx_nreasyworkspace_tcemain.php:&tx_nreasyworkspace_tcemain->lockedIcon';
?>

class.tx_nreasyworkspace_tcemain.php
<?php
declare(encoding='UTF-8');
/**
 * Netresearch easy workspace extension for Typo3
 *
 * PHP version 5
 *
 * @category Typo3
 * @package  nr_easy_workspace
 * @author   Alexander Opitz <alexander.opitz@netresearch.de>
 * @author   Christian Weiske <christian.weiske@netresearch.de>
 * @license  http://www.gnu.org/licenses/gpl.txt GPL v3
 * @link     http://mantis.nr/view.php?id=4424
 */

/**
 * Hooks for Typo3 TCE and pagetree
 *
 * @category Typo3
 * @package  nr_easy_workspace
 * @author   Alexander Opitz <alexander.opitz@netresearch.de>
 * @author   Christian Weiske <christian.weiske@netresearch.de>
 * @license  http://www.gnu.org/licenses/gpl.txt GPL v3
 * @link     http://mantis.nr/view.php?id=4424
 */
class tx_nreasyworkspace_tcemain
{
    /**
    * Debugging on or off.
    * When activated, logging to devlog is active.
    *
    * @var boolean
    */
    private $debug = false;



    /**
     * This method is called by a hook in the TYPO3 Core Engine (TCEmain).
     *
     * @param string $command    ???
     * @param string $table      The table TCEmain is currently processing
     * @param string $id         The records id (if any)
     * @param mixed  $value      ???
     * @param object &$reference Reference to the parent object (TCEmain)
     *
     * @return void
     */
    public function processCmdmap_postProcess(
        $command, $table, $id, $value, &$reference
    ) {
        if ($this->debug) {
            t3lib_div::devLog(
                'processCmdmap_postProcess', 'nr_easy_workspace', 0,
                array($command, $table, $id, $value)
            );
        }

        switch ($command) {
        case 'version':
            switch ((string)$value['action']) {
            case 'live_page':
                $this->live_page($table, $id, $reference->cmdmap, $reference);
                break;
            case 'live_page_down':
                $this->live_page($table, $id, $reference->cmdmap, $reference, true);
                break;
            }
            break;
        }
    }//function processCmdmap_postProcess(..)



    /**
    * Push the given record into live workspace
    *
    * @param string  $table      Name of table that is being processed
    * @param integer $id         uid of record being processed
    * @param string  $strCmd     Command map from &$reference
    * @param object  &$reference Reference to parent object (TCEmain)
    * @param boolean $recursive  Process child records, too
    *
    * @return void
    */
    protected function live_page(
        $table, $id, $strCmd, &$reference, $recursive = false
    ) {
        global $TCA, $BE_USER;

        $wsid = $BE_USER->workspace;
        // Nicht live schalten sonder reviewn?
        if (isset($strCmd['pages'][0]['version']['review'])) {
            $this->review_page($table, $id, $strCmd, $reference);
            return;
        }

        $curVersion = t3lib_BEfunc::getRecord($table, $id, '*');
        $offVersion = t3lib_BEfunc::getWorkspaceVersionOfRecord(
            $wsid, $table, $id, '*'
        );

        if (is_array($offVersion) && $id !== $offVersion['uid']) {
            $reference->version_swap($table, $id, $offVersion['uid']);
        } else {
            $this->setT3verTo(0, $table, $id);
        }

        if ($table != 'pages') {
            return;
        }

        $arInPage = $this->selectVersionsInWorkspace(
            $wsid, $filter = 0, $stage = -99, $id
        );
        foreach ($arInPage as $onTable => $arElements) {
            if ($onTable == 'pages') {
                if ($recursive) {
                    foreach ($arElements as $arElement) {
                        $this->live_page(
                            'pages', $arElement['uid'], $strCmd, $reference,
                            $recursive
                        );
                    }
                } else {
                    continue;
                }
            } else {
                foreach ($arElements as $arElement) {
                    $reference->version_swap(
                        $onTable, $arElement['t3ver_oid'], $arElement['uid']
                    );
                }
            }
        }
        if ($recursive) {
            $arPages = t3lib_BEfunc::getRecordsByField('pages', 'pid', $id);
            if (is_array($arPages)) {
                foreach ($arPages as $arPage) {
                    $this->live_page(
                        'pages', $arPage['uid'], $strCmd, $reference, $recursive
                    );
                }
            }
        }
    }//protected function live_page(..)



    /**
    * Send page to reviewer
    * "Setzen der Seiten auf Reviewstatus, sowie eintragen in zu sendende
    *  Mail Liste-"
    *
    * @param string  $table      Name of table that is being processed
    * @param integer $id         uid of record being processed
    * @param string  $strCmd     Command map from &$reference
    * @param object  &$reference Reference to parent object (TCEmain)
    * @param boolean $recursive  Process child records, too
    *
    * @return void
    */
    protected function review_page(
        $table, $id, $strCmd, &$reference, $recursive = false
    ) {
        global $BE_USER, $TYPO3_CONF_VARS, $LANG;

        $strPathLL = 'EXT:nr_easy_workspace/examples';

        if (isset($TYPO3_CONF_VARS['EXT']['extConf']['nr_easy_workspace'])) {
            $arConfig = unserialize(
                $TYPO3_CONF_VARS['EXT']['extConf']['nr_easy_workspace']
            );
            if (isset($arConfig['locallangDirectory'])
                && $arConfig['locallangDirectory'] !== ''
            ) {
                $strPathLL = $arConfig['locallangDirectory'];
            }
        }

        $strLLFile = t3lib_div::getFileAbsFileName($strPathLL) . '/locallang.xml';
        if (is_file($strLLFile)) {
            $this->messageLL = $LANG->includeLLFile($strLLFile, 0);
        } else {
            $this->messageLL = $GLOBALS['LOCAL_LANG'];
        }

        //kick out people without edit access
        if (!$reference->checkRecordUpdateAccess($table, $id)) {
            $reference->newlog(
                'Attempt to set stage for record failed'
                . ' because you do not have edit access',
                1
            );
            return;
        }



        $wsid = $BE_USER->workspace;

        $curVersion = t3lib_BEfunc::getRecord($table, $id, '*');
        $offVersion = t3lib_BEfunc::getWorkspaceVersionOfRecord(
            $wsid, $table, $id, '*'
        );

        $stat = $reference->BE_USER->checkWorkspaceCurrent();

        $bHasAccess = t3lib_div::inList(
            'admin,online,offline,reviewer,owner',
            $stat['_ACCESS']
        );
        if (!$bHasAccess && !($stageId <= 1 && $stat['_ACCESS']==='member')) {
            $reference->newlog(
                'The member user tried to set a stage value "'
                . $stageId . '" which was not allowed',
                1
            );
            return;
        }

        $arConfVars
            =& $TYPO3_CONF_VARS['SC_OPTIONS']['tx_nreasyworkspace_tcemain'];

        $arSendMails = $arConfVars['sendMails'];
        // Set stage of record:

        if (is_array($offVersion)) {
            $this->setT3verTo(1, $table, $offVersion['uid']);
        } else {
            $this->setT3verTo(1, $table, $id);
        }

        switch ($table) {
        case 'pages':
            $pageId = $id;
            // Set the elements inside the page to review
            $arInPage = $this->selectVersionsInWorkspace(
                $wsid, 0, -99, $pageId
            );
            foreach ($arInPage as $table => $arElements) {
                foreach ($arElements as $arElement) {
                    $this->setT3verTo(1, $table, $arElement['uid']);
                }
            }
            break;
        case 'pages_language_overlay':
            $pageId = $curVersion['pid'];
            $arGetVars['L'] = $curVersion['sys_language_uid'];
            break;
        }

        // Set change into logfile
        $reference->newlog(
            'Stage for record was changed to ' . $stageId
            . '. Comment was: "' . substr($comment, 0, 100) . '"'
        );
        //TEMPORARY, except 6-30 as action/detail number which is
        // observed elsewhere!
        $reference->log(
            $table, $id, 6, 0, 0, 'Stage raised...',
            30, array('comment' => $comment, 'stage' => $stageId)
        );

        $emails = $reference->notifyStageChange_getEmails(
            $strCmd['pages'][0]['version']['reviewer'], true
        );
        $emails = array_unique($emails);

            // Send email:
        if (count($emails)) {
            $message = sprintf(
                "%s\n\n%s",
                ($strCmd['pages'][0]['version']['msg_ind'] != ''
                    ? $strCmd['pages'][0]['version']['msg_ind']
                    : $LANG->getLLL(
                        'mailtext_' . $strCmd['pages'][0]['version']['msg'],
                        $this->messageLL
                    )
                ),
                "%s" // Späteres ersetzen durch alle zu reviewenden Seiten
            );
            $arSendMails['to']      = $emails;
            $arSendMails['message'] = $message;
            $arSendMails['title']   = ($strCmd['pages'][0]['version']['msg'] != ''
                ? $LANG->getLLL(
                        'mailtext_' . $strCmd['pages'][0]['version']['msg'] . '_header',
                        $this->messageLL
                    )
                : 'TYPO3 Workspace Note: Please review pages'
            );
            $arSendMails['toreplace'] .= "\n"
                . $this->getPreviewUrl($pageId, $arGetVars, $anchor);
        }
        $arConfVars['sendMails'] = $arSendMails;
    }//protected function review_page(..)

    /**
     * Generates preview url with domain record for page.
     *
     * @param integer $id         Id of page to link
     * @param array   $arGetVars  Possible get vars to add to url.
     * @param string  $anchor     Possible anchor to add to url.
     *
     * @return string Link to preview of page.
     */
    public function getPreviewUrl($id, array $arGetVars = null, $anchor = '')
    {
        global $BE_USER;
        $rootLine = t3lib_BEfunc::BEgetRootLine($id);

        $strPreview  = TYPO3_mainDir
            . 'mod/user/ws/wsol_preview.php?id=' . $id
            . '&wsid=' . $BE_USER->workspace;

        if ($rootLine)  {
            $parts = parse_url(t3lib_div::getIndpEnv('TYPO3_SITE_URL'));
            if (t3lib_BEfunc::getDomainStartPage($parts['host'], $parts['path'])) {
                $preUrl_temp = t3lib_BEfunc::firstDomainRecord($rootLine);
            }
        }
        $preUrl = $preUrl_temp
            ? (t3lib_div::getIndpEnv('TYPO3_SSL') ? 'https://' : 'http://')
            . $preUrl_temp . '/'
            : t3lib_div::getIndpEnv('TYPO3_SITE_URL');

        $strGetVars = is_array($arGetVars)
            ? '&' .http_build_query($arGetVars)
            : '';

        return $preUrl . $strPreview . $strGetVars . $anchor;
    } // public function getPreviewUrl(..)

    /**
    * Changes the stage version (t3ver_stage) of elements in database.
    *
    * Stages:
    * -1 rejected
    *  0 editing
    *  1 review
    * 10 publish
    *
    * @param integer $nStage Staging version (t3ver_stage)
    * @param string  $table  Table to do that on
    * @param integer $id     uid to change stage on
    *
    * @return void
    */
    protected function setT3verTo($nStage, $table, $id) {
        $sArray = array(
            't3ver_stage' => $nStage
        );
        $GLOBALS['TYPO3_DB']->exec_UPDATEquery(
            $table, 'uid=' . intval($id), $sArray
        );
    }

    /**
    * Selectieren aller Elemente in einem Workspace, von einer
    * ID aus gehend über alle Tabellen
    *
    * @param integer $wsid   ID of current workspace, see $BE_USER->workspace
    * @param integer $filter ???
    * @param integer $stage  ???
    * @param integer $pageId ID of page to get element versions from
    *
    * @return array Array of elements in the given workspace and the given
    *               page id
    */
    function selectVersionsInWorkspace(
        $wsid, $filter = 0, $stage = -99, $pageId = -1
    ) {
        global $TCA;

        $wsid   = intval($wsid);
        $filter = intval($filter);
        $output = array();

        // Traversing all tables supporting versioning:
        foreach ($TCA as $table => $cfg) {
            if ($TCA[$table]['ctrl']['versioningWS']) {
                //Select all records from this table in the database
                // from the workspace
                //This joins the online version with the offline version
                // as tables A and B
                $output[$table] = $GLOBALS['TYPO3_DB']->exec_SELECTgetRows(
                    'A.uid, A.t3ver_oid,'
                    . ($table === 'pages'
                        ? ' A.t3ver_swapmode,'
                        :''
                    )
                    . ' B.pid AS realpid',
                    $table . ' A,' . $table.' B',
                    //Table A is the offline version and pid=-1 defines offline
                    'A.pid=-1'
                    . ($pageId!=-1
                        ? ($table==='pages'
                            ? ' AND B.uid='.intval($pageId)
                            : ' AND B.pid='.intval($pageId)
                        )
                        : ''
                    )
                    //For "real" workspace numbers, select by that.
                    //If = -98, select all that are NOT online (zero).
                    //Anything else below -1 will not select on the wsid
                    // and therefore select all!
                    . ($wsid > -98
                        ? ' AND A.t3ver_wsid=' . $wsid
                        : ($wsid === -98
                            ? ' AND A.t3ver_wsid!=0'
                            : ''
                        )
                    )
                    //lifecycle filter:
                    // 1 = select all drafts (never-published),
                    // 2 = select all published one or more times (archive/multiple)
                    . ($filter===1
                        ? ' AND A.t3ver_count=0'
                        : ($filter === 2
                            ? ' AND A.t3ver_count>0'
                            : ''
                        )
                    )
                    . ($stage!=-99
                        ? ' AND A.t3ver_stage=' . intval($stage)
                        : ''
                    )
                    //Table B (online) must have PID >= 0 to signify being online.
                    . ' AND B.pid>=0'
                    //... and finally the join between the two tables.
                    . ' AND A.t3ver_oid=B.uid'
                    . t3lib_BEfunc::deleteClause($table, 'A')
                    . t3lib_BEfunc::deleteClause($table, 'B'),
                    '',
                    //Order by UID, mostly to have a sorting in the
                    // backend overview module which doesn't "jump around"
                    // when swapping.
                    'B.uid'
                );
            }
        }

        return $output;
    }//function selectVersionsInWorkspace(..)



    /**
    * Add an extra symbol to the page in the page tree in the backend
    * when the page is under review.
    *
    * @param array       $arRecordInfo Array with table name and record uid,
    *                                  e.g. array('pages', 12)
    * @param webPageTree &$tree        Tree object that is being generated
    *
    * @return string Additional data that shall be displayed in the tree item
    */
    public function lockedIcon($arRecordInfo, &$tree)
    {
        global $BE_USER, $LANG;

        if ($BE_USER->getTSConfigVal('treeview_showreviewstate') === 'none') {
            return '';
        }
        list($strTable, $nUid) = $arRecordInfo;
        if ($strTable != 'pages') {
            return '';
        }
        //t3lib_BEfunc::getRecordWSOL($table, $uid, $fields = '*', $where = '',)
        $arRecord = t3lib_BEfunc::getRecordWSOL(
            $strTable, $nUid,
            'uid,t3ver_stage,t3ver_wsid'
        );

        if ($arRecord['t3ver_wsid'] == 0) {
            return '';
        }
        
        switch((int)$arRecord['t3ver_stage']) {
            case 0:
                $strImage = $this->icons('edit', $LANG->getLL('stage_editing'));
                break;
            case 1:
                $strImage = $this->icons('review', $LANG->getLL('label_review'));
                break;
            case 10:
                $strImage = $this->icons('publish', $LANG->getLL('label_publish'));
                break;
            case -1:
                $strImage = $this->icons('rejected', $LANG->getLL('label_rejected'));
                break;
            default:
                $strImage = $this->icons('undefined', $LANG->getLL('label_undefined'));
                break;
        }
        //add icon to tell the page is under review
        return $strImage;
//            '<img src="/typo3/gfx/lightning.png" alt="Review" title="In Review"/>';
    }//public function lockedIcon(..)

    function icons($type, $strTitle='') {
        switch($type)   {
            case 'edit':
                $icon   = 'gfx/edit2.gif';
                $width  = 11;
                $height = 12;
            break;
            case 'review':
                $icon = 'gfx/lightning.png';
                $width  = 16;
                $height = 16;
            break;
            case 'publish':
                $icon = 'gfx/icon_ok2.gif';
                $width  = 18;
                $height = 16;
            break;
            case 'rejected':
                $icon = 'gfx/icon_warning.gif';
                $width  = 18;
                $height = 16;
            break;
            case 'undefined':
                $icon = 'gfx/icon_fatalerror.gif';
                $width  = 18;
                $height = 16;
            break;
        }
        if ($icon)  {
            return '<img'.
                t3lib_iconWorks::skinImg(
                    $this->doc->backPath,
                    $icon,
                    'width="'.$width.'" height="'.$height.'"'
                )
                .' class="absmiddle"'
                .' title="'.$strTitle.'"'
                .' alt="'.$strTitle.'"'
                .' />';
        }
    }
}//class tx_nreasyworkspace_tcemain


$file = 'ext/nr_easy_workspace/class.tx_nreasyworkspace_tcemain.php';
if (defined('TYPO3_MODE') && $TYPO3_CONF_VARS[TYPO3_MODE]['XCLASS'][$file]) {
    include_once $TYPO3_CONF_VARS[TYPO3_MODE]['XCLASS'][$file];
}

?>